ngcc: implement a new program-based entry-point finder

petebacondarwin opened the pull request:
Add a new `EntryPointFinder` that can be seeded from the imports in the program specified by a tsconfig.json file. This should be much faster than the `DirectoryWalkerEntryPointFinder` when the active program only imports a small proportion of the installed entry-points.

On the AIO project only half the entry-points would be processed by ngcc if the CLI async integration with ngcc made use of the new flag.

petebacondarwin commented:
The approach here, which involves parsing each source file, could be slow for a large app.
We should investigate using the TS scanner directly or even a simple regular expression, since we only care about capturing the import paths and can be quite tolerant of false positives.

JoostK commented on packages/compiler-cli/ngcc/test/entry_point_finder/program_based_entry_point_finder_spec.ts:
```diff
@@ -0,0 +1,147 @@
+    });
+
+    describe('findEntryPoints()', () => {
+      it('should find entry-points imported into the program', () => {
```
I would like to see a test for the scenario where path mappings into TS source files are used, as that is done in some projects (e.g. [this article](https://blog.angulartraining.com/path-mapping-with-typescript-cac16d12c416) and [another one](https://dev.to/cooperdev/restructure-with-ease-thanks-to-typescript-path-mappings-4b0e)).

As these are absolute imports, ngcc's module resolver will probably recognize them as such and not traverse into the source files, causing parts of the program not to be scanned. I haven't verified this assumption, so could be wrong here.

petebacondarwin replied:
Actually ngcc's module resolver understands pathMappings and so would indeed resolve these. I will add to the test...

petebacondarwin replied:
Added

JoostK reviewed (commented):
Typo in "capture dynamic import expressions as well as declarations" commit:

- "and creates lesss" -> "and creates less"

petebacondarwin commented on packages/compiler-cli/ngcc/src/command_line_options.ts:
```diff
@@ -116,6 +123,7 @@ export function parseCommandLineOptions(args: string[]): NgccOptions {
   const enableI18nLegacyMessageIdFormat = options['legacy-message-ids'];
   const invalidateEntryPointManifest = options['invalidate-entry-point-manifest'];
   const errorOnFailedEntryPoint = options['error-on-failed-entry-point'];
+  const findEntryPointsFromTsConfigProgram = options['use-tsconfig'];
```
😱

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/esm_dependency_host.ts:
```diff
@@ -13,20 +13,200 @@ import {DependencyHostBase} from './dependency_host';
+    return this.tryStringLiteral();
+  }
+
+  protected skipStarClause(): ts.SyntaxKind|null {
```
Strictly they have two names depending upon whether they are in an `import` or an `export` statement, which is why I didn't go with that name. But given that it was weird enough for you to comment, I will change it.

petebacondarwin replied:
I'm going with `skipNamespacedClause()` and `skipNamedClause()`.

JoostK commented on packages/compiler-cli/ngcc/src/dependencies/esm_dependency_host.ts:
```diff
@@ -13,20 +13,200 @@ import {DependencyHostBase} from './dependency_host';
+      }
+      lastToken = currentToken;
+    }
+    return imports;
```
Just thought of this: consider resetting the scanner's text to the empty string to avoid holding on to the most recent file content.

petebacondarwin replied:
Sounds reasonable.

petebacondarwin replied:
Done here https://github.com/angular/angular/pull/37075/commits/ffc717a68be19616f77676b207c3196cdb21e95a

petebacondarwin committed:
refactor(ngcc): move shared code into `DependencyHostBase`

The various dependency hosts had a lot of duplicated code.
This commit refactors them to move this into the base class.

petebacondarwin committed:
fixup! fix(ngcc): ensure that more dependencies are found by `EsmDependencyHost`

petebacondarwin committed:
fix(ngcc): capture dynamic import expressions as well as declarations

Previously we only checked for static import declaration statements.
This commit also finds import paths from dynamic import expressions.

Also this commit should speed up processing: Previously we were parsing
the source code contents into a `ts.SourceFile` and then walking the parsed
AST to find import paths.
Generating an AST is unnecessary work and it is faster and creates less
memory pressure to just scan the source code contents with the TypeScript
scanner, identifying import paths from the tokens.

petebacondarwin committed:
fix(ngcc): ensure that more dependencies are found by `EsmDependencyHost`

Previously this host was skipping files if they had imports that spanned
multiple lines, or if the import was a dynamic import expression.

petebacondarwin committed:
fixup! fix(ngcc): capture dynamic import expressions as well as declarations

petebacondarwin committed:
fixup! feat(ngcc): implement a program-based entry-point finder

petebacondarwin committed:
feat(ngcc): implement a program-based entry-point finder

This finder is designed to only process entry-points that are reachable
by the program defined by a tsconfig.json file.

It is triggered by calling `mainNgcc()` with the `findEntryPointsFromTsConfigProgram`
option set to true. It is ignored if a `targetEntryPointPath` has been
provided as well.

It is triggered from the command line by adding the `--use-program-dependencies`
option, which is also ignored if the `--target` option has been provided.

Using this option can speed up processing in cases where there is a large
number of dependencies installed but only a small proportion of the
entry-points are actually imported into the application.

petebacondarwin committed:
fixup! fix(ngcc): capture dynamic import expressions as well as declarations

gkalpak reviewed (approved):
Reviewed-for: dev-infra

gkalpak commented on packages/compiler-cli/ngcc/src/dependencies/esm_dependency_host.ts:
```diff
@@ -93,7 +225,7 @@ export class EsmDependencyHost extends DependencyHostBase {
  */
 export function hasImportOrReexportStatements(source: string): boolean {
-  return /(import|export)\s.+from/.test(source);
+  return /(?:import|export)[\s\S]+?(["'])(?:(?:\\\1|.)*?)\1/.test(source);
```
Nice :+1:

There is one unneeded set of parenthesis (afaict), but not really important:
```suggestion
  return /(?:import|export)[\s\S]+?(["'])(?:\\\1|.)+?\1/.test(source);
```

petebacondarwin replied:
I'm a bit frightened to touch that again :-)

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/tracing_entry_point_finder.ts:
```diff
@@ -0,0 +1,206 @@
+ * This is faster than searching the entire file-system for all the entry-points,
+ * and is used primarily by the CLI integration.
+ *
+ * There are two concrete implementation of this class.
```
```suggestion
 * There are two concrete implementations of this class.
```

gkalpak reviewed (commented):
Post-merge review :grin::
**tl;dr** LGTM :+1:

petebacondarwin commented:
Thanks for the extra review @gkalpak - here is a fixup #37461
